Bugfix/onboarding auto name - Using Futures

n13 opened the pull request:
## Futures vs Bloc
@pdahlberg I couldn't quite get the Bloc to behave the way I wanted to so after spending a huge amount of time on that I switched to futures.

It might be a better fit here since we're not actually having a stream of data.

Mainly the stream was just too obscure. Future is very simple, and yet there's still a lot of complexity

## Design goals
- The Seeds account name page starts out with a valid account name - if the first name derived from the full name is not available, it chooses another one. 
- This means the first choice will always be an available and valid one - the user will never see an invalid account show up on this page
- If they hit create account they are good to go
- if they change the account, and it is not following the rules, I fade in the rules instead of the button which is disabled anyway
- If they change the account to something that is following the rules, but is not available, an error is shown ("not available"). 

In this last case, it would make sense to show alternative, available names. For simplicity. However, our created names are not very beautiful - just replacing the last character with a 1.

So I figured:
- If a user didn't care about their name, then they could just use the first one and be done
- If they do want to choose their name, chances are they won't like to use their desired name with a 1 at the end, so they can try to find another name in the text field

For this case I think we could better assist users and make more intelligent suggestions, using words, phrases, their full name. But I want to ship this now... ;) 

TODO: Verify on larger / normal screen devices. Design could probably loosen up on larger devices. Animations popping up and down should be nicer and use scrollable view

n13 committed:
intermediary checkin with new state and new screen

n13 committed:
fixed lint issues

n13 committed:
user name screen finished

n13 committed:
updated test

n13 committed:
almost working version with Stream

n13 committed:
remove push hooks - i often push draft versions

n13 committed:
WIP initial version

works but does not yet have replacement options. Some refactoring

n13 committed:
added hypha server

n13 committed:
slightly better layout on small devices

n13 committed:
optional inputFormatters added

n13 committed:
lint

n13 committed:
close to final version - no suggestions yet but very usable

n13 committed:
version

needs to be larger than iOS app store

pdahlberg commented on lib/features/account/create_account_bloc.dart:
```diff
@@ -24,22 +24,22 @@ class CreateAccountBloc {
     _initGenerateAccountFromUserAccount();
-    _initValidateLatestAccountUnlessAlreadyValid();
+    _initGenerateAccountFromUserName();
+    //_initValidateLatestAccountUnlessAlreadyValid();
```
Remove commented out code

n13 replied:
Didn't even mean to touch this file... since I didn't end up using bloc

n13 replied:
Done

pdahlberg commented on lib/screens/onboarding/create_account_account_name.dart:
```diff
@@ -0,0 +1,259 @@
+    }
+  }
+
+  String initialUsername(String nickname) {
```
I would like to see this logic moved out of the UI

n13 replied:
Fair point - no real reason it's in there, could be in the service

Normally I would make extensions on strings that do these things - an EOS string extension that does everything related to EOS string modification. We can talk about this.

One thing that stumped me - and maybe I don't understand something there - is that these services mostly require a context so I can access them.

But when a widget is initialized I need to set some initial values and there's no context available.

Setting initial state shouldn't be a complicated process... ?! Probably missing something here.

pdahlberg replied:
Do you mean to get the instance from Provider? 
You have access to the context in the overridable initState() where I think init code is sensible to run. 
You can of course also supply values in the constructor.

pdahlberg commented on lib/screens/onboarding/create_account_account_name.dart:
```diff
@@ -0,0 +1,259 @@
+              errorString = "$currentText is not availale";
+            }
+
+            var createEnabled = valid && definitelyAvailableOnChain;
```
Could this (line 69-88) be put somewhere else?

n13 replied:
Not sure what you mean

pdahlberg commented on .githooks/pre-push:
```diff
@@ -1,11 +0,0 @@
-#!/usr/bin/env bash
```
Why did you remove these?

n13 replied:
They kept getting in my way trying to push code that wasn't ready for production

I like to check in drafts 

I would much rather these being caught on github when I am done with the code - then it can go through and do this linting. But not when I want to push - it's taking very long, then it complains even though the code isn't even finished. 

For safety I like to do more checkins and pushes. 

Not a huge deal but it was a case where the tools consistently got in my way

Tools are supposed to help, not get in the way ;)

pdahlberg reviewed (commented):
Hmm... I do get that the BLOC pattern and Streams can be a bit opaque before you can wrap your head around it. Though I disagree with the "not having a stream of data". Seems to me that the code that was natural to put into the Bloc now ends up in the UI class _CreateAccountAccountNameState instead. The build method starts with a lot of logic to derive which state the app actually is in and switches between different UI bits depending on this derived state. 

Using Futures can absolutely be an alternative to BLOC, but in that case I would really like to see all that code to figure out state put into some other class. For example the service class. That way it can be tested as well as not polluting the UI code. 

I really would like to get BLOCs working in a way that everyone is comfortable with though because as far as I've seen they do keep the UI building very clean which is something I really like. That said, I'm totally open to using some other pattern or state framework if you or anyone else got suggestions. 

I'm sure the code works so I'm happy to publish this if you really want to get it out quickly, but would you mind if I gave this a go with BLOC at the same time? Or before you publish is also an option of course.

n13 committed:
remove commented out code

n13 committed:
more error messages

re-ordered error messages so "less than 12" is one of the last in case users enter illegal character they get notified

n13 committed:
Revert "remove push hooks - i often push draft versions"

This reverts commit eaaaf32a381c087b845b6444141b6a8932976cf0.

pdahlberg reviewed (approved):
Almost forgot to add the approval so you can merge it.
